test/spmd/test_tracing.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import torch
+
+from torch.distributed._spmd import CommTensor
+from torch.distributed.distributed_c10d import (
+    get_global_rank,
+    get_world_size,
+)
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.testing._internal.common_utils import run_tests
+from spmd.testing.common_utils import (  # type: ignore
+    DistTensorTestBase,
+    with_comms,
+)
+from spmd.tensor import DeviceMesh
+
+from typing import List
+
+
+class TraceDeviceMeshTest(DistTensorTestBase):
+    @property
+    def world_size(self) -> int:
+        return 8
+
+    @with_comms
+    def test_tracing_all_reduce_nd(self):
+        mesh_tensor = torch.arange(8).reshape(2, 2, 2)
+        mesh = DeviceMesh(self.device_type, mesh_tensor)
+        local_tensor = torch.ones(3, 3, device=self.device_type) * self.rank
+
+        # check all dim groups
+        dim_to_subgroups = mesh.get_dim_groups()
+        for dim, dim_group in enumerate(dim_to_subgroups):
+            dim_group_size = get_world_size(dim_group)
+            global_ranks = [
+                get_global_rank(dim_group, i) for i in range(dim_group_size)
+            ]
+
+            def fn(tensor: torch.Tensor):
+                comm_tensor = CommTensor(tensor)
+                reduced_tensor = mesh.all_reduce(comm_tensor, mesh_dim=dim)
```

Comment:
is there a reason we are calling the collective under the `for` loop over dim_groups?